Feature: Added Talawa-admin to run system daemon service in Linux, Added Eslint for Markdown file and updated the deprecated eslint

sahitya-chandra opened the pull request:
**What kind of change does this PR introduce?**
feature

**Issue Number:**

Fixes #2595 

**Did you add tests for your changes?**
No

**Snapshots/Videos:**

Working proof of talawa_admin.service file
[Screencast from 2025-01-03 13-10-02.webm](https://github.com/user-attachments/assets/0df5d616-4135-445e-b6c1-55cf90396cce)

**If relevant, did you update the documentation?**

No

**Summary**

- [x]  Created examples/linux/systemd folder in the root directory

- [x]  Inside examples/linux/systemd added talawa_admin.service file

- [x]  Inside examples/linux folder added Daemon_Service_Setup.md file for proper setup of talawa-admin service

**Does this PR introduce a breaking change?**

No

**Other information**

I have locally tested all the changes that I have made to the codebase

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**

Yes

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
	- Added systemd service configuration for Talawa-Admin.
	- Introduced comprehensive Linux service setup guide in installation documentation.

- **Documentation**
	- Updated installation instructions with detailed systemd service setup steps.
	- Provided troubleshooting guidance for service management.

- **Chores**
	- Modified ESLint configuration to adjust file linting rules, including the addition of the markdown plugin.
	- Removed `.eslintignore` file and updated linting patterns.
	- Updated `devDependencies` by removing `@testing-library/dom` and adding `eslint-plugin-markdown`.
	- Consolidated import statements in test files for improved readability.
	- Updated command for generating HTML reports for code coverage.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

palisadoes commented:
1. Add the examples directory to the repo's eslint configuration file
1. Move the content of the markdown file to the root directory INSTALLATION.md for
1. Commit and push to your upstream. The error should go away.

There is no need to close the PR

sahitya-chandra commented:
> 1. Add the examples directory to the repo's eslint configuration file
> 2. Move the content of the markdown file to the root directory INSTALLATION.md for
> 3. Commit and push to your upstream. The error should go away.
> 
> There is no need to close the PR

@palisadoes Sir Can I put examples directory in .eslintignore file

palisadoes commented:
1. The .eslintignore file is deprecated https://eslint.org/docs/latest/use/configure/ignore-deprecated
1. Update the replacement with the appropriate contents .eslintignore and delete .eslintignore

sahitya-chandra commented:
@palisadoes Sir I have made the changes. You can review them now.

Cioppolo14 commented:
@sahitya-chandra Please fix the failed tests and ensure Coderabbit approves your PR.

sahitya-chandra commented:
@Cioppolo14 eslint workflow will always fail beacuse .md and .service file will not get linted by eslint
and @palisadoes sir instructed me to delete .eslintignore file that's why `sensitive files have been changed without authorization` workflow is failing

sahitya-chandra commented on INSTALLATION.md:
```diff
@@ -380,3 +390,72 @@ If you don't want this hook to run, you can manually opt out of this using the `
+
+### Service Configuration
+
+#### 1. **Copy the `talawa_admin.service` file**
+   - Place the `talawa_admin.service` file in the appropriate systemd directory based on your Linux distribution:
+     - For most distributions: `/etc/systemd/system/`
+     - For systems using `systemd`, this will be the default directory.
+
+#### 2. **Verify the CODEROOT Path**
+   - Ensure that the `CODEROOT` environment variable matches the absolute path to the Talawa-Admin code directory.
+
+#### 3. **Set the Correct Working Directory**
+   - Always use the absolute path for the `WorkingDirectory`. Do **not** use `$CODEROOT` in the `WorkingDirectory` field.
+
+#### 4. **Ensure the `.env` File Exists**
+   - Verify that the path in the `EnvironmentFile` line points to a valid `.env` file located in the root directory of the Talawa-Admin repository.
+
```
I have made these changes in latest commit

Cioppolo14 commented:
@sahitya-chandra I believe I follow.  What about the skipped tests? Why were they skipped?

sahitya-chandra commented:
@Cioppolo14 sir the code change that I have made include adding `talawa_admin.service` file because of this linting failed. Eslint was unable to parse this `talawa_admin.service` file.

And the second one failed because `.eslintignore` file is deprecated currently 
So I deleted it and added the files that should be ignored in `eslint.config.json` file.
And it is a sensetive file.

sahitya-chandra commented:
@palisadoes sir what else I can do...

noman2002 commented on .eslintignore:
```diff
@@ -1,8 +0,0 @@
-# Contains the PDF file of the Tag as JSON string, thus does not need to be linted
-src/components/CheckIn/tagTemplate.ts
-package.json
-package-lock.json
-tsconfig.json
-
-# Ignore the Docusaurus website subdirectory
-docs/**
```
Why we need to remove this file ?

sahitya-chandra replied:
@noman2002 sir acc. to new eslint version docs their is no need of `.eslintignore` file in the codebase we can directly add the files or folder that we want to ignore in `.eslintrc.json` file.
So @palisadoes sir told me to delete `.eslintignore` file and update `.eslintrc.json` file

sahitya-chandra replied:
![Screenshot from 2025-01-05 17-03-15](https://github.com/user-attachments/assets/37dcb6a5-4062-40cb-9b11-39dafc4e767e)
and this is working correctly

palisadoes commented:
Please fix the failing tests

palisadoes commented on examples/linux/systemd/talawa_admin.service:
```diff
@@ -0,0 +1,53 @@
+Environment=NODE_ENV=production
+
+# Absolute path is needed for working directory
+WorkingDirectory=/path/to/your/talawa-admin
```
Please look at the example in the link I provided to see how variables are used later in the file.

Update this to use the variable CODEROOT and move this below the "No Need" line

sahitya-chandra replied:
We cannot use any variable in WorkingDirectory. We have to use absolute path otherwise the file will not work.

sahitya-chandra replied:
If you want we can remove the WorkingDirectory completely then the file will work fine but some logs will not be visible

palisadoes commented on examples/linux/systemd/talawa_admin.service:
```diff
@@ -0,0 +1,53 @@
+# Absolute path is needed for working directory
+WorkingDirectory=/path/to/your/talawa-admin
+
+EnvironmentFile=/path/to/your/talawa-admin/.env
```
Please look at the example in the link I provided to see how variables are used later in the file.

Update this to use the variable CODEROOT and move this below the "No Need" line

sahitya-chandra replied:
I have modified the EnvironmentFile.

palisadoes reviewed (changes requested):
See comment

palisadoes commented:
Please fix the failing tests

sahitya-chandra commented:
> Please fix the failing tests
@palisadoes sir only this eslint test is failing
![Screenshot from 2025-01-05 17-24-09](https://github.com/user-attachments/assets/99ff570e-8f8c-45c2-8438-d56785d6b4dd)

is their any way to fix this

palisadoes commented:
Add it as a linting exception

palisadoes commented:
1. The error is in an invalid character in INSTALLATION.md
2. Please fix it

sahitya-chandra commented:
@palisadoes sir the line that is causing error is Line no. 1 and I have not even touched it

sahitya-chandra commented on examples/linux/systemd/talawa_admin.service:
```diff
@@ -0,0 +1,52 @@
+################################################################################
+
+EnvironmentFile=$CODEROOT/.env
+ExecStart=/bin/bash -c "source /path/to/your/.nvm/nvm.sh && nvm use default && npm run serve"
```
done

sahitya-chandra commented:
@palisadoes I think that if you run `npx eslint INSTALLION.md` in your repo it will always fail.
This is the only reason that this eslint check fails.

palisadoes commented:
1. Why isn't it failing for anyone else?
2. Please merge with the latest upstream. 
3. Fix the file

palisadoes commented:
Fix the conflicting file

sahitya-chandra commented:
@palisadoes sir I think that it will be nice if we create a separate markdown file for this service.

sahitya-chandra commented:
and It is not failing for others because they have not made any change to this specific file that is failing test

palisadoes commented:
1. Fix the file.
2. It's failing because your markdown is invalid.
3. Use a markdown viewer to see the source of the issue.
4. I can see at least one case of invalid syntax
5. Run a linter against it before committing to the PR

sahitya-chandra commented:
@palisadoes I have removed all the changes from this file that I have made and then I have run the linter 
It is still failing.
And when I run the linter for upstream It failed again

sahitya-chandra commented:
@palisadoes @Cioppolo14 
1. So, I have found that eslint for markdown file is not currently present in the repo and when some run `npx eslint any-file.md ` the cmd fails.
This issue is present in all the branches including main branch.

2. And because the eslint test CI workflow only runs on the files that have been changed, it fails for my this PR, 
as their is no linter for markdown file and I have made changes to INSTALLATION.md file.

3. This test does not fails for everyone, it only fails for those who have made changes to any markdown file.

In my next commit I have added eslint-plugin-markdown for `.md` files.
Please have a look.

sahitya-chandra commented:
@palisadoes @Cioppolo14 
Every test has now passed...
Please review.

sahitya-chandra commented:
@palisadoes sir please review

sahitya-chandra commented:
@palisadoes sir please review

noman2002 commented:
@sahitya-chandra Please resolve the conflicts.

palisadoes commented:
Please fix the failing tests

noman2002 commented:
@sahitya-chandra Any update on this ?

sahitya-chandra commented:
@noman2002 everything is working fine.
But someone has added new CI workflow and it is failing again and again.
I don't know what should I do now ?

noman2002 commented:
Please investigate why its failing. Also resolve the conflict.

sahitya-chandra commented:
@noman2002 ok sir .
I will check if i can fix it.

sahitya-chandra commented:
@coderabbitai full review

sahitya-chandra commented:
@coderabbitai full review

sahitya-chandra commented:
@coderabbitai full review

sahitya-chandra commented:
@coderabbitai full review

sahitya-chandra commented:
@coderabbitai full review

sahitya-chandra commented:
@noman2002 sir all checks passed.
Please review.

palisadoes commented:
Closing.

Please resubmit a PR where these files are excluded.

- src/screens/UserPortal/Volunteer/VolunteerManagement.spec.tsx
- package.json
- package-lock.json
- CONTRIBUTING.md
- .eslintignore
- .eslintrc.json


Your changes to .eslintrc.json will severely limit the stability of our code. They are not in keeping with the needs of the community or the end user
